plugins/scripting/scripting-compiler-impl/src/org/jetbrains/kotlin/scripting/extensions/ScriptingResolveExtension.kt

```diff
@@ -22,8 +26,17 @@ class ScriptingResolveExtension : SyntheticResolveExtension {
         declarationProvider: PackageMemberDeclarationProvider,
         result: MutableSet<ClassDescriptor>
     ) {
+        val previousImplicitReceivers = hashSetOf<KotlinType>()
+        for (file in declarationProvider.getPackageFiles()) {
```

Comment:
This looks too expensive for me and probably incorrect too. 
What you probably should do instead is to collect descriptors compiling each line (see history handling or places where `earlierScriptsForReplInterpreter` is set) and get receivers from the script descriptors and pass them to the newly created descriptors.

Reply:
Ok, I will try to refactor this logic

Reply:
BTW, how should I get the script descriptor(s) here? By `ctx` or using more explicit way?